File: projects/mcp-8004-enterprise/agents_implementation/hf-document-server.js
/**
 * hf-document-server.js
 *
 * MCP server for the HedgeFundDocumentAgent card.
 * Implements tools from agents/mcp/hf-document.mcp.json.
 */

import { ethers }        from 'ethers';
import { startMcpServer } from './mcp-server-base.js';

const tools = {

  assemble_documents: {
    description: 'Assemble client documents in response to a bank data request.',
    inputSchema: {
      type: 'object',
      properties: {
        flow_id:     { type: 'string' },
        request_id:  { type: 'string' },
        oracle_type: { type: 'string', enum: ['aml', 'credit'] },
        spec_hash:   { type: 'string' },
        round:       { type: 'integer' },
        trace_id:    { type: 'string' },
      },
      required: ['flow_id', 'request_id', 'oracle_type', 'spec_hash'],
    },
    handler({ flow_id, oracle_type, spec_hash, round, trace_id }) {
      if (!flow_id) throw new Error('flow_id required');
      console.log(`[hf-document-server] [${trace_id ?? 'n/a'}] assemble_documents flow=${flow_id} oracle=${oracle_type} round=${round ?? 1}`);

      const docTypes = oracle_type === 'aml'
        ? ['passport_copies', 'corporate_registry', 'beneficial_ownership', 'source_of_funds']
        : ['audited_financials', 'nav_statements', 'credit_references', 'prime_broker_letter'];

      const dataHash = ethers.keccak256(ethers.toUtf8Bytes(`hf-docs:${flow_id}:${oracle_type}:${spec_hash}:round${round ?? 1}`));
      return { data_hash: dataHash, documents: docTypes };
    },
  },
};

startMcpServer({ tools, resources: [], prompts: {}, serverLabel: 'hf-document-server' });

File: projects/mcp-8004-enterprise/agents_implementation_py/servers/hf_document_server.py
"""
hf_document_server.py — MCP server for the hedge fund document agent.
Port of hf-document-server.js with LCEL chain.

Implements tools from agents/mcp/hf-document.mcp.json.

Usage:
    python -m servers.hf_document_server [port]   # default 8020
"""

from __future__ import annotations

import json
import logging
import os
import sys
from pathlib import Path
from typing import Literal, Optional

from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from pydantic import BaseModel
from web3 import Web3

from servers.server_base import create_server, suspended_when_revoked
from shared.bounds_monitor_client import report_tool_call
from shared.prompt_hash import load_langchain_messages

logger = logging.getLogger(__name__)

# ── Load MCP spec & build LCEL chain ─────────────────────────────────────────

_SPEC_PATH = (
    Path(__file__).parent.parent.parent / "agents" / "mcp" / "hf-document.mcp.json"
)
_spec = json.loads(_SPEC_PATH.read_text(encoding="utf-8"))
_lc_messages = load_langchain_messages(_spec)

_prompt = ChatPromptTemplate.from_messages(
    [(msg["role"], msg["content"]) for msg in _lc_messages]
)


class _DocDecision(BaseModel):
    documents: list[str] = ["corporate_structure", "beneficial_ownership", "financials", "source_of_funds"]
    notes: str = ""


_chain = None


def _get_chain():
    global _chain
    if _chain is None:
        llm = ChatOpenAI(
            model=os.getenv("OPENAI_MODEL", "gpt-4o-mini"),
            temperature=0,
        ).with_structured_output(_DocDecision)
        _chain = _prompt | llm
    return _chain


def _keccak(text: str) -> str:
    return Web3.keccak(text=text).hex()


mcp = create_server("hf-document-server")


@mcp.tool()
@suspended_when_revoked("assemble_documents")
async def assemble_documents(
    flow_id: str,
    request_id: str,
    oracle_type: str,
    spec_hash: str,
    round: int = 1,
    trace_id: str = "",
) -> dict:
    """Assemble client documents in response to a bank data request."""
    logger.info(
        "[hf-doc-server] [%s] assemble_documents flow=%s oracle=%s round=%s",
        trace_id or "n/a", flow_id, oracle_type, round,
    )

    decision: _DocDecision = await _get_chain().ainvoke(
        {"oracle_type": oracle_type, "flow_id": flow_id}
    )
    report_tool_call("assemble_documents", success=True)

    data_hash = _keccak(f"hf-docs:{flow_id}:{oracle_type}:round{round}")
    return {
        "data_hash": data_hash,
        "documents": decision.documents,
    }


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8020
    logger.info("[hf-document-server] Starting on port %d", port)
    mcp.run(transport="streamable-http", host="0.0.0.0", port=port)
